tools/regtesting/do_regtest.py

```diff
@@ -0,0 +1,437 @@
+    print("\n------------------------------------------------------------------------")
+    print("Copying test files...", end="")
+    shutil.copytree(cfg.cp2k_root / "tests", cfg.work_base_dir)
+    print("done")
+
+    # Read TEST_TYPES.
+    test_types_fn = Path(cfg.cp2k_root / "tests" / "TEST_TYPES")
+    test_types: List[Optional[TestType]] = [None]  # test type zero
+    lines = test_types_fn.read_text(encoding="utf8").split("\n")
+    test_types += [TestType(l) for l in lines[1 : int(lines[0]) + 1]]
+
+    # Read TEST_DIRS.
+    batches: List[Batch] = []
+    test_dirs_fn = Path(cfg.cp2k_root / "tests" / "TEST_DIRS")
+    for line in test_dirs_fn.read_text(encoding="utf8").split("\n"):
+        line = line.split("#", 1)[0].strip()
+        if not line:
+            continue
+        batch = Batch(line, cfg)
+
+        # Read TEST_FILES.
+        test_files_fn = Path(batch.src_dir / "TEST_FILES")
+        for line in test_files_fn.read_text(encoding="utf8").split("\n"):
+            line = line.split("#", 1)[0].strip()
+            if not line:
+                continue
+            batch.tests.append(Test(line, test_types, batch.workdir))
+        batches.append(batch)
+
+    # TODO: Add unit tests.
+    # TODO: Reorder test dirs.
+    # TODO: Cancel when there are >100 failures.
+
+    # Create async tasks.
+    tasks = []
+    num_restrictdirs = 0
+    for batch in batches:
+        if not batch.requirements_satisfied(flags, cfg.mpiranks):
+            print(f"Skipping {batch.name} because its requirements are not satisfied.")
+        elif not any([re.match(p, batch.name) for p in cfg.restrictdirs]):
+            num_restrictdirs += 1
+        elif batch.name in ALLWAYS_SKIP_DIRS:
+            print(f"Skipping {batch.name} because it doesn't work with the cp2k shell.")
+        else:
+            tasks.append(asyncio.create_task(process_batch(batch, cfg)))  # launch
+
+    if num_restrictdirs:
+        print(f"Skipping {num_restrictdirs} test directories because of --restrictdir.")
+    if not tasks:
+        print("\nNot test directories selected, check --restrictdir filter.")
+        sys.exit(1)
+
+    # Wait for tasks to finish and print their results.
+    print(f"Launched {len(tasks)} test directories and {cfg.num_workers} worker...\n")
+    all_results: List[TestResult] = []
+    with open(cfg.error_summary, "wt", encoding="utf8", errors="replace") as err_fh:
+        for num_done, task in enumerate(asyncio.as_completed(tasks)):
+            batch_result = await task
+            all_results += batch_result.results
+            print(f">>> {batch_result.batch.workdir}")
+            print("\n".join([str(r) for r in batch_result.results]))
+            print(f"<<< {batch_result.batch.workdir} ({num_done + 1}", end="")
+            print(f" of {len(tasks)}) done in {batch_result.duration:.2f} sec")
+            sys.stdout.flush()
+            err_fh.write("\n".join([r.error for r in batch_result.results if r.error]))
```

Comment:
there should be no need to for the explicit list comprehension here

Reply:
Indeed. Thanks!